### PR TITLE
Fix resource clean-up script

### PR DESCRIPTION
**Pull request by weshaggard**
Fixed "rg.Name" to "rg.ResourceGroupName" and added some extra logging.

### CODE REVIEW

**Review by benbp — APPROVED** [review]

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 1]
```diff
@@ -19,18 +19,33 @@ function Get-PurgeableGroupResources {
 }
 
 function Get-PurgeableResources {
+    $purgeableResources = @()
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:22]: You don't need to collect them to then return them. Anything written to the pipeline is returned. If only 0 or 1 is, that's why I wrap the function in a `@()` call which coerces to a collection. Why the change here?

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 2]
```diff
@@ -19,18 +19,33 @@ function Get-PurgeableGroupResources {
...
+    Write-Verbose "Retrieving the purgeable resources from subscription $subscriptionId"
+
+    Write-Verbose "Retrieving deleted Key Vaults from subscription $subscriptionId"
+
     # Get deleted Key Vaults for the current subscription.
-    Get-AzKeyVault -InRemovedState `
+    $deletedKeyVaults  = Get-AzKeyVault -InRemovedState `
         | Add-Member -MemberType NoteProperty -Name AzsdkResourceType -Value 'Key Vault' -PassThru
 
+    Write-Verbose "Found $($deletedKeyVaults.Count) deleted Key Vaults to potentiall purge."
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:33]:
```suggestion
    Write-Verbose "Found $($deletedKeyVaults.Count) deleted Key Vaults to potentially purge."
```

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 3]
```diff
@@ -40,7 +55,14 @@ function Get-PurgeableResources {
                 EnablePurgeProtection = $r.properties.purgeProtectionEnabled
             }
         }
+        Write-Verbose "Found $($deletedHsms.Count) deleted Managed HSM Key Vaults to potentially purge."
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:58]: Managed HSMs are not Key Vaults
```suggestion
        Write-Verbose "Found $($deletedHsms.Count) deleted Managed HSMs to potentially purge."
```

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 4]
```diff
@@ -40,7 +55,14 @@ function Get-PurgeableResources {
...
+        $purgeableResources += $deletedHsms
+    }
+    else {
+        Write-Verbose "Found no deleted HSMs to purge."
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:62]: Key Vault also uses HSMs. "Managed HSMs" are different.
```suggestion
        Write-Verbose "Found no deleted Managed HSMs to purge."
```

File: eng/scripts/live-test-resource-cleanup.ps1 [hunk 5]
```diff
@@ -73,9 +74,13 @@ foreach ($rg in $toDelete)
 {
   if ($Force -or $PSCmdlet.ShouldProcess("$($rg.ResourceGroupName) (UTC: $($rg.Tags.DeleteAfter))", "Delete Group")) {
     # Add purgeable resources that will be deleted with the resource group to the collection.
-    $purgeableResources += Get-PurgeableGroupResources $rg.Name
+    $purgeableResourcesFromRG = Get-PurgeableGroupResources $rg.ResourceGroupName
 
-    Write-Verbose "Deleting group: $($rg.Name)"
+    if ($purgeableResourcesFromRG -and $purgeableResourcesFromRG.Count -gt 0) {
```
Comment (heaths) [inline eng/scripts/live-test-resource-cleanup.ps1:79]: Nit: The count is unnecessary. PowerShell is truthy, which means an undefined collection variable == null collection == an empty collection.

**Review by heaths — CHANGES_REQUESTED** [review]

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 6]
```diff
@@ -19,18 +19,33 @@ function Get-PurgeableGroupResources {
...
+    $purgeableResources = @()
     $subscriptionId = (Get-AzContext).Subscription.Id
 
+    Write-Verbose "Retrieving the purgeable resources from subscription $subscriptionId"
+
+    Write-Verbose "Retrieving deleted Key Vaults from subscription $subscriptionId"
+
     # Get deleted Key Vaults for the current subscription.
-    Get-AzKeyVault -InRemovedState `
+    $deletedKeyVaults  = Get-AzKeyVault -InRemovedState `
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:30]: Given the error in the pipeline, add `-ErrorAction Ignore` to both this `Get-AzKeyVault` call and the one above. If the RG was deleted between getting the RG to delete and now, this errs. Using `Ignore` will just ignore the error and not emit any value.

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 (hunk 1 above)
Comment (heaths) [reply eng/common/scripts/Helpers/Resource-Helpers.ps1:22]: Or was the idea merely to make sure you return a collection and not a collection of collections? You could do this, or just use `Write-Output $mycollection` which, by default, writes each item in the collection to the pipeline.

The advantage of doing this is that any subsequent cmdlets that handles pipeline output can start processing right away instead of waiting for all output.

```powershell
function foo { $a = 1..5; write-output $a; write-host "break"; $b = 6..10; write-output $b }
filter bar { write-host "Got $_" } # equivalent to function bar -process {}
foo | bar
```
```
Got 1
Got 2
Got 3
Got 4
Got 5
break
Got 6
Got 7
Got 8
Got 9
Got 10
```

Comment (heaths) [reply eng/common/scripts/Helpers/Resource-Helpers.ps1:22]: And if you'd still rather collect into a collection then return, may as well remove the "and coerce into ..." comments and get rid of the `@()` wrapper around `Get-PurgeableResources`. At the very least, you'd still be returning an empty collection from this PR's changes.

Still the checks below of `$myCollectionVar -and $myCollectionVar.Count` are unnecessary. Just `$myCollectionVar` is sufficient.

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 7]
```diff
@@ -1,123 +1,136 @@
...
+    [string] $ResourceGroupName
+  )
+  $purgeableResources = @()
+
+  Write-Verbose "Retrieving deleted Key Vaults from resource group $ResourceGroupName"
+
+  # Get any Key Vaults that will be deleted so they can be purged later if soft delete is enabled.
+  $deletedKeyVaults = Get-AzKeyVault -ResourceGroupName $ResourceGroupName -ErrorAction Ignore | ForEach-Object {
+    # Enumerating vaults from a resource group does not return all properties we required.
+    Get-AzKeyVault -VaultName $_.VaultName | Where-Object { $_.EnableSoftDelete } `
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:15]: I wonder if the `-ErrorAction Ignore` should be here as well in the off chance - but possible - even it was deleted between these REST calls. I checked and it still terminates the pipe on this line (i.e. so `where-object` doesn't complain about an empty pipe) so it would be safe.

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 8]
```diff
@@ -1,123 +1,136 @@
...
+      | Add-Member -MemberType NoteProperty -Name AzsdkResourceType -Value 'Key Vault' -PassThru
+  }
+
+  if ($deletedKeyVaults) {
+    Write-Verbose "Found $($deletedKeyVaults.Count) deleted Key Vaults to potentially purge."
+    $purgeableResources += $deletedKeyVaults
+  }
+
+  Write-Verbose "Retrieving deleted Managed HSM Key Vaults from resource group $ResourceGroupName"
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:24]:
```suggestion
  Write-Verbose "Retrieving deleted Managed HSMs from resource group $ResourceGroupName"
```

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 9]
```diff
@@ -1,123 +1,136 @@
...
+
+  # Get any Managed HSMs in the resource group, for which soft delete cannot be disabled.
+  $deletedHsms = Get-AzKeyVaultManagedHsm -ResourceGroupName $ResourceGroupName -ErrorAction Ignore `
+    | Add-Member -MemberType NoteProperty -Name AzsdkResourceType -Value 'Managed HSM' -PassThru
+
+  if ($deletedHsms) {
+    Write-Verbose "Found $($deletedHsms.Count) deleted Managed HSM Key Vaults to potentially purge."
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:31]:
```suggestion
    Write-Verbose "Found $($deletedHsms.Count) deleted Managed HSMs to potentially purge."
```

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 10]
```diff
@@ -1,101 +1,136 @@
...
+
+  if ($deletedKeyVaults) {
+    Write-Verbose "Found $($deletedKeyVaults.Count) deleted Key Vaults to potentially purge."
+    $purgeableResources += $deletedKeyVaults
+  }
+
+  Write-Verbose "Retrieving deleted Managed HSM Key Vaults from resource group $ResourceGroupName"
+
+  # Get any Managed HSMs in the resource group, for which soft delete cannot be disabled.
+  $deletedHsms = Get-AzKeyVaultManagedHsm -ResourceGroupName $ResourceGroupName -ErrorAction Ignore `
```
Comment (weshaggard) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:27]: @heaths do we need to filter this to ones that have EnableSoftDelete as well?

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 11]
```diff
@@ -1,123 +1,136 @@
...
+  # Get deleted Key Vaults for the current subscription.
+  $deletedKeyVaults  = Get-AzKeyVault -InRemovedState `
+    | Add-Member -MemberType NoteProperty -Name AzsdkResourceType -Value 'Key Vault' -PassThru
+
+  if ($deletedKeyVaults) {
+    Write-Verbose "Found $($deletedKeyVaults.Count) deleted Key Vaults to potentially purge."
+    $purgeableResources += $deletedKeyVaults
+  }
+
+  Write-Verbose "Retrieving deleted Managed HSM Key Vaults from subscription $subscriptionId"
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:52]:
```suggestion
  Write-Verbose "Retrieving deleted Managed HSMs from subscription $subscriptionId"
```

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 [hunk 12]
```diff
@@ -1,123 +1,136 @@
...
-                EnablePurgeProtection = $r.properties.purgeProtectionEnabled
-            }
-        }
-        Write-Verbose "Found $($deletedHsms.Count) deleted Managed HSM Key Vaults to potentially purge."
-        $purgeableResources += $deletedHsms
-    }
-    else {
-        Write-Verbose "Found no deleted HSMs to purge."
+    if ($deletedHsms) {
+      Write-Verbose "Found $($deletedHsms.Count) deleted Managed HSM Key Vaults to potentially purge."
```
Comment (heaths) [inline eng/common/scripts/Helpers/Resource-Helpers.ps1:73]:
```suggestion
      Write-Verbose "Found $($deletedHsms.Count) deleted Managed HSMs to potentially purge."
```

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 (hunk 10 above)
Comment (heaths) [reply eng/common/scripts/Helpers/Resource-Helpers.ps1:27]: No. Managed HSMs always have soft delete enabled, just like new Key Vaults. It cannot be turned off, so checking for `EnableSoftDelete` is redundant. I also double-checked with the service team.

**Review by heaths — APPROVED** [review]

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 (hunk 1 above)
Comment (weshaggard) [reply eng/common/scripts/Helpers/Resource-Helpers.ps1]: removed the coerce operators.

File: eng/common/scripts/Helpers/Resource-Helpers.ps1 (hunk 7 above)
Comment (weshaggard) [reply eng/common/scripts/Helpers/Resource-Helpers.ps1:15]: I will add it here as well.
